### PR TITLE
Prefab Overrides: Fix property-edit issues in DPE Entity Inspector

### PR DESCRIPTION
**Pull request by forkercat**
## What does this PR do?

_Feature is behind two flags: `ed_enableDPE` and `EnableInspectorOverrideManagement`_

This PR fixes couple of issues after we changed to adopt DPE patches in https://github.com/o3de/o3de/pull/14988.

Inspector Override RFC: https://github.com/o3de/sig-content/issues/112

**Change List:**
1. Fixes https://github.com/o3de/o3de/issues/15241
2. Fixes https://github.com/o3de/o3de/issues/15242
3. Fixes https://github.com/o3de/o3de/issues/15246 (this PR adds add-element support)
4. Adds logic of skipping update in undo node if property is not changed.
5. Improves handling of default override values:
    1. Override patch is created if property is changed to non-default value and won't go away even users change it back to default value.
    2. Override patch won't be generated if property is not changed to other value from default value.
6. Vector-type property is now stored as a vector rather than individual values as shown below.

```json
{
    "op": "replace",
    "path": "/Entities/Entity_[123]/Components/Component_[8896558300816993180]/Transform Data/Translate",
    "value": [0, 3, 0]
}
```

**Issues Not Fixed:**
1. Undo/redo and revert is not tested. It should be addressed in a different PR. See GHI: https://github.com/o3de/o3de/issues/15243
1. For dynamic container list, 'delete' does not work. From DPE, we are receiving index paths pointing at individual values. Still investigating how to make DPE to generate path to the whole value.

https://user-images.githubusercontent.com/11157226/226967528-043ea7c5-7136-435c-b5fe-2b0d6549fc93.mp4

## How was this PR tested?

Manually tested in editor. Will have a follow-up PR to add unit tests, but it should be done after undo/redo issues are fixed.

Here are some workflows I tested and verified patches stored in file.

**Translate vector in TransformComponent:**
1. Change value from [0, 0, 0], to [0, 0, 0] (no icon shows up)
1. Change value from [0, 0, 0], to [3, 0, 0] (icon shows up), to [3, 5, 0] (icon shows up), to [0, 5, 0] (icon shows up), to [0, 0, 0] (icon still shows up)
1. Change value from [0, 0, 0] to [3, 0, 0], then to [3, 0, 0]. The last edit is not valid. Undoing once will change value back to [0, 0, 0].
1. Focus on the owning prefab and change the value to [3, 0, 0]. Exit focus. Change to [5, 0, 0] (icon shows up), to [3, 0, 0] (icon still shows up).
1. Tested 'Tab' to complete editing. It works as expected.

**Container list in TagComponent:**
1. Tested adding new elements.
1. Tested updating values in new elements.

Also tested switch, slider, color picker, text, enum type properties.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoComponentPropertyOverride.cpp [hunk 1]
```diff
@@ -73,62 +74,77 @@ namespace AzToolsFramework
...
-
-            m_overriddenPropertyPathFromFocusedPrefab /= relativePathFromOwningPrefab;
-
-            const TemplateId topTemplateId = link->get().GetTargetTemplateId();
-            const PrefabDom& topTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom(topTemplateId);
+            // Configure paths from focused and top templates. The top template is the source template of the link.
+            AZ::Dom::Path pathToPropertyFromTop =
+                AZ::Dom::Path(PrefabInstanceUtils::GetRelativePathFromClimbedInstances(climbUpResult.m_climbedInstances, true)) /
+                relativePathFromOwningPrefab;
+            AZ::Dom::Path pathToPropertyFromFocused(PrefabDomUtils::InstancesName);
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoComponentPropertyOverride.cpp:81]: nit: pathToPropertyFromFocusedPrefab to be clearer

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoComponentPropertyOverride.cpp [hunk 2]
```diff
@@ -73,62 +74,77 @@ namespace AzToolsFramework
...
+            pathToPropertyFromFocused /= climbUpResult.m_climbedInstances.back()->GetInstanceAlias(); // top instance alias
+            pathToPropertyFromFocused /= pathToPropertyFromTop;
+
+            // Get the current state of property value from focused template.
+            const TemplateId focusedTemplateId = link->get().GetTargetTemplateId();
+            const PrefabDom& focusedTemplateDom = m_prefabSystemComponentInterface->FindTemplateDom(focusedTemplateId);
+            PrefabDomPath domPathToPropertyFromFocused(pathToPropertyFromFocused.ToString().c_str());
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoComponentPropertyOverride.cpp:88]: same: domPathToPropertyFromFocusedPrefab

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoComponentPropertyOverride.cpp [hunk 3]
```diff
@@ -73,62 +74,77 @@ namespace AzToolsFramework
...
-                // DOM value pointers can't be relied upon if the original DOM gets modified after pointer creation.
-                PrefabDomPath overriddenPropertyDomPath(pathFromFocusedPrefab.ToString().c_str());
-                const PrefabDomValue* overriddenPropertyDomInTopTemplate = overriddenPropertyDomPath.Get(topTemplateDom);
-
-                
-                PrefabDom overridePatches;
-                if (overriddenPropertyDomInTopTemplate)
+                // Get the default state of property value from top template (direct child prefab of focused template).
+                // Note: We can't get the default state from the owning template because top template might have stored overrides
+                // on the owning instance. Those overrides should be treated as default/base values from the focused template.
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoComponentPropertyOverride.cpp:107]: Do you mean to say , 'we can't get the default state from the "focused template"' ? The comment is not clear imo and can be removed altogether as I don't see it clarifying things.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoUtils.h [hunk 4]
```diff
@@ -17,6 +17,12 @@ namespace AzToolsFramework
     {
         namespace PrefabUndoUtils
         {
+            enum class PatchOperationType
+            {
+                Add,
+                Replace
+            };
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoUtils.h:24]: why not reuse https://github.com/o3de/o3de/blob/development/Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Overrides/PrefabOverrideTypes.h#L15-L20 ?

**Review by srikappa-amzn — COMMENTED** [review]
Looking great! Quite a few issues fixed, which is nice!

Unit tests in the future for this will be crucial to prevent regressions as clearly there are several intricate edge cases here.

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoComponentPropertyOverride.cpp (hunk 3 above)
Comment (forkercat) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoComponentPropertyOverride.cpp:107]: I think I can update it. I want to put a note here to address reader's question like "why can't we just use the source template from owning instance as the default state?". The reason is there could be the case that `topInstance != owningInstance` and topInstance has an override edit to change the default state from source.

**Comment by forkercat** [comment]
Might need to merge from dev to resolve conflicts. Will run AR later.

**Review by srikappa-amzn — APPROVED** [review]

**Review by AMZN-daimini — APPROVED** [review]
